### PR TITLE
Scripted pipeline

### PR DESCRIPTION
**Pull request by BulatSaif**
Now this pipeline should have all feature that old pipeline had.

### Description of the Change
Rewrite declarative pipeline to scripted pipeline

### Benefits
Flexibly

### Possible Drawbacks 
May appear Blue Ocean visualization problems.

### CODE REVIEW

File: .jenkinsci/builders/x64-linux-build-steps.groovy [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+def successPostSteps(scmVars, String build_type, List environment) {
+  stage('successPostSteps') {
+    withEnv(environment) {
+      artifacts = load ".jenkinsci/artifacts.groovy"
+      filesToUpload = []
+      platform = sh(script: 'uname -m', returnStdout: true).trim()
+      if (build_type == 'Release' && scmVars.GIT_LOCAL_BRANCH ==~ /(master|develop|dev)/) {
+        sh("mkdir -p build/artifacts")
+        sh("mv ./build/iroha.deb ./build/iroha.tar.gz build/artifacts")
```
Comment (bakhtin) [inline .jenkinsci/builders/x64-linux-build-steps.groovy]: Either `.deb` or `.tar.gz` may be missing depending on the cmake options set. `mv` would throw an error in that case

File: .jenkinsci/builders/x64-linux-build-steps.groovy [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}
+
+def successPostSteps(scmVars, String build_type, List environment) {
+  stage('successPostSteps') {
+    withEnv(environment) {
+      artifacts = load ".jenkinsci/artifacts.groovy"
+      filesToUpload = []
+      platform = sh(script: 'uname -m', returnStdout: true).trim()
+      if (build_type == 'Release' && scmVars.GIT_LOCAL_BRANCH ==~ /(master|develop|dev)/) {
```
Comment (bakhtin) [inline .jenkinsci/builders/x64-linux-build-steps.groovy]: I would have moved all the logic related to checking branch names, release types and other cases when we need to decide on whether do something or not to Jenkinsfile. Previously there were separate builds for each case (e.g. `x64ReleaseLinuxBuild`) to separate general functions from different build cases described here: https://docs.google.com/spreadsheets/d/1fCt1EPakDI-aBWE17UsZe7_GjTqhaDI6iskuStYb6ok/

File: .jenkinsci/builders/x64-mac-build-steps.groovy [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    buildDir = 'build'
+    compilers = vars.compilerMapping()
+    cmakeBooleanOption = [ (true): 'ON', (false): 'OFF' ]
+        cmakeBuildOptions = ""
+
+    if (packagebuild){
+      cmakeBuildOptions = " --target package "
+    }
+    //not sure if we need this
+    //env.IROHA_VERSION = "0x${scmVars.GIT_COMMIT}"
```
Comment (bakhtin) [inline .jenkinsci/builders/x64-mac-build-steps.groovy]: We do not need this

File: .jenkinsci/builders/x64-mac-build-steps.groovy [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    //env.IROHA_HOME = "/opt/iroha"
+    //env.IROHA_BUILD = "${env.IROHA_HOME}/build"
+
+    utils.ccacheSetup(5)
+
+    for (compiler in compilerVersions.split(',')) {
+      stage ("build ${compiler}"){
```
Comment (bakhtin) [inline .jenkinsci/builders/x64-mac-build-steps.groovy]: See comment above about arranging `stage` block and `for` loop

File: .jenkinsci/utils/doxygen.groovy [hunk 5]
```diff
@@ -1,12 +1,12 @@
 #!/usr/bin/env groovy
 
 def doDoxygen() {
+  sh "doxygen Doxyfile"
   // TODO: Remove this comment once dev branch will return to develop
   // I will not be changing branches here. It requires some rewriting
   // Hope dev branch situation will be resolved soon
   if (env.GIT_LOCAL_BRANCH in ["master","develop"] || env.CHANGE_BRANCH_LOCAL == 'develop') {
```
Comment (bakhtin) [inline .jenkinsci/utils/doxygen.groovy]: Should be refactored to work with dev branch. But, again, I would recommend to move these checks to Jenkinsfile

File: .jenkinsci/builders/x64-linux-build-steps.groovy [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    --network=${env.IROHA_NETWORK} postgres:9.5 -c 'max_prepared_transactions=100'"
+    iC.inside(""
+    + " -e IROHA_POSTGRES_HOST=${env.IROHA_POSTGRES_HOST}"
+    + " -e IROHA_POSTGRES_PORT=${env.IROHA_POSTGRES_PORT}"
+    + " -e IROHA_POSTGRES_USER=${env.IROHA_POSTGRES_USER}"
+    + " -e IROHA_POSTGRES_PASSWORD=${env.IROHA_POSTGRES_PASSWORD}"
+    + " --network=${env.IROHA_NETWORK}"
+    + " -v /var/jenkins/ccache:${env.CCACHE_DEBUG_DIR}") {
+      utils.ccacheSetup(5)
+      for (compiler in compilerVersions.split(',')) {
```
[the remark this answers is not in the document]
Comment (BulatSaif) [reply .jenkinsci/builders/x64-linux-build-steps.groovy]: I will move code to successPostSteps(), artifacts will be moved once build is completed

File: .jenkinsci/builders/x64-linux-build-steps.groovy (hunk 1 above)
Comment (BulatSaif) [reply .jenkinsci/builders/x64-linux-build-steps.groovy]: Flag `packagebuild` responsible for building both packages, if one of them is missing it is should handle as error.
```
            -DPACKAGE_DEB=${cmakeBooleanOption[packagebuild]} \
            -DPACKAGE_TGZ=${cmakeBooleanOption[packagebuild]}
```

File: .jenkinsci/builders/x64-linux-build-steps.groovy [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+      }
+      stage("Analysis") {
+            cppcheck ? build.cppCheck(buildDir, parallelism) : echo('Skipping Cppcheck...')
+            sonar ? build.sonarScanner(scmVars, environment) : echo('Skipping Sonar Scanner...')
+      }
+      stage('Build docs'){
+        docs ? doxygen.doDoxygen() : echo("Skipping Doxygen...")
+      }
+      stage ('DockerManifestPush'){
+        if (pushDockerTag) {
```
[the remark this answers is not in the document]
Comment (BulatSaif) [reply .jenkinsci/builders/x64-linux-build-steps.groovy]: now it pushed if `specialBranch` flag is enabled
```
  if (scmVars.GIT_LOCAL_BRANCH in ["master","develop","dev"] || scmVars.CHANGE_BRANCH_LOCAL in ["develop","dev"]) {
    specialBranch =  true
  }else {
    specialBranch = false
  }
```

File: .jenkinsci/builders/x64-linux-build-steps.groovy [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+          utils.dockerPush(iC, "${platform}-develop-build")
+          DockerManifestPush(iC, "develop-build", environment)
+          if (build_type == 'Release' && scmVars.GIT_LOCAL_BRANCH ==~ /(master|develop|dev)/){
+            sh "cp ./build/iroha.deb docker/release/iroha.deb"
+            iCRelease = docker.build("${env.DOCKER_REGISTRY_BASENAME}:${scmVars.GIT_COMMIT}-${env.BUILD_NUMBER}-release", "--no-cache -f docker/release/Dockerfile ${WORKSPACE}/docker/release")
```
[the remark this answers is not in the document]
Comment (BulatSaif) [reply .jenkinsci/builders/x64-linux-build-steps.groovy]: you mean move `dockerManifestPush` function inside of `utils.groovy` or `docker-pull-or-build.groovy`?

**Comment by bakhtin** [comment]
Please, make a test run of the pipeline to check the upload of Docker images and other artifacts. For that replace `dev` and `master` branch instances in the code with the name of the current branch. Also make sure you push Docker images with some other tag (not `latest` or `develop` or `develop-build`)

**Comment by BulatSaif** [comment]
1. There were some conflict in  load functions(`vars =  load ".jenkinsci/vars.groovy"`), using `def` in front helps.
2. `mac-mcst-3 `have different curl version that break upload process,  using ` curl --http1.1` helps 
3. There are several conflict in build parameter, 
ex: 
 -`fuzzing` and `compiler version`,  `fuzzing` works only with `clang`
-`tests_` and build type `Release`, tests do not build, 
-`package` and build type  `Debug`,  *.tar.gz empty , tests empty 
-`sanitize` and   build type `Release`
-etc
so checkbox becomes useless, you can not select that you want.
there are to solutions that I see:
1.rewrite the cmakeFile  
2.reduce options, see: https://soramitsu.atlassian.net/browse/DOPS-202

@bakhtin package are  pushed  successfully.
- please, delete image with tag `*not-supposed-to-be-pushed`
https://hub.docker.com/r/hyperledger/iroha/tags 
- please, folders` PR-1960-20181220*`
https://nexus.iroha.tech/#browse/browse:artifacts:iroha%2Fmacos
https://nexus.iroha.tech/#browse/browse:artifacts:iroha%2Flinux%2Fx86_64

**Pull request merged**